src/main/kotlin/models/EngineModel.kt

```diff
@@ -62,23 +73,38 @@ class EngineModel(
     }
 
     fun transform(transformation: ImageProcessing) {
-        val previous = if (currIndex < 0) originalImage.value else snapshots[currIndex]
-
-        transformations.subList(currIndex + 1, transformations.size).clear()
-        snapshots.subList(currIndex + 1, snapshots.size).clear()
+        transform(transformation, "preview")
+    }
 
-        snapshots.add(
-            WritableImage(
-                previous.pixelReader,
-                previous.width.toInt(),
-                previous.height.toInt()
-            )
-        )
-        transformations.add(transformation)
-        currIndex++
-        updateListSelection()
-        transformation.process(snapshots[currIndex])
-        previewImage.value = snapshots[currIndex]
+    /* the @param destination here refers to where to put the transformed image: the image panel or the decode panel */
+    fun transform(transformation: ImageProcessing, destination: String) {
+        val previous = if (currIndex < 0) originalImage.value else snapshots[currIndex]
+        when (destination) {
+            "preview" -> {
+                transformations.subList(currIndex + 1, transformations.size).clear()
+                snapshots.subList(currIndex + 1, snapshots.size).clear()
+
+                snapshots.add(
+                    WritableImage(
+                        previous.pixelReader,
+                        previous.width.toInt(),
+                        previous.height.toInt()
+                    )
+                )
+                transformations.add(transformation)
+                currIndex++
+                updateListSelection()
+                transformation.process(snapshots[currIndex])
+                previewImage.value = snapshots[currIndex]
+            }
+            "decode" -> {
+                if (transformation is SteganographyDecoder) {
+                    val decoder: SteganographyDecoder = transformation
+                    transformation.process(previous as WritableImage)
```

Comment:
When click on decode, this line cause 'image cannot be cast to writable image exception'